tests/ContextTest.php

```diff
@@ -2,68 +2,152 @@
+        $context = new Context(['foo' => 'bar']);
+        $context->setData(['bar' => 'foo']);
 
-        $context->mergeUserData(['foo' => 'bar']);
-        $context->mergeUserData(['baz' => 'bar']);
+        $this->assertAttributeEquals(['foo' => 'bar', 'bar' => 'foo'], 'data', $context);
 
-        $this->assertEquals(['foo' => 'bar', 'baz' => 'bar'], $context->getUserData());
+        $context->setData(['foo' => ['bar' => 'baz']]);
+
+        $this->assertAttributeEquals(['foo' => ['bar' => 'baz'], 'bar' => 'foo'], 'data', $context);
     }
 
-    public function testMergeUserDataWithSameKey()
+    public function testReplaceData()
     {
-        $context = new Context();
+        $context = new Context(['foo' => 'bar']);
+        $context->replaceData(['bar' => 'foo']);
 
-        $context->mergeUserData(['foo' => 'bar']);
-        $context->mergeUserData(['foo' => 'baz']);
+        $this->assertAttributeEquals(['bar' => 'foo'], 'data', $context);
+    }
+
+    public function testClear()
+    {
+        $context = new Context(['foo' => 'bar']);
+
+        $this->assertAttributeEquals(['foo' => 'bar'], 'data', $context);
+
+        $context->clear();
 
-        $this->assertEquals(['foo' => 'baz'], $context->getUserData());
+        $this->assertAttributeEmpty('data', $context);
     }
 
-    public function testSetUserData()
+    public function testGetDataByPath()
+    {
+        $context = new Context(['foo' => ['bar' => 'baz']]);
+
+        $this->assertEquals('baz', $context->getDataByPath('[foo][bar]'));
+    }
+
+    public function testSetDataByPath()
     {
         $context = new Context();
+        $context->setDataByPath('[foo]', ['bar' => []]);
+
+        $this->assertAttributeEquals(['foo' => ['bar' => []]], 'data', $context);
 
-        $context->setUserData(['foo' => 'bar']);
-        $context->setUserData(['bar' => 'baz']);
+        $context->setDataByPath('[foo][bar]', 'baz');
 
-        $this->assertEquals(['bar' => 'baz'], $context->getUserData());
+        $this->assertAttributeEquals(['foo' => ['bar' => 'baz']], 'data', $context);
     }
 
-    public function testMergeExtraData()
+    public function testIsEmpty()
     {
         $context = new Context();
 
-        $context->mergeExtraData(['foo' => 'bar']);
-        $context->mergeExtraData(['baz' => 'bar']);
+        $this->assertTrue($context->isEmpty());
 
-        $this->assertEquals(['foo' => 'bar', 'baz' => 'bar'], $context->getExtraData());
+        $context->setData(['foo' => 'bar']);
+
+        $this->assertFalse($context->isEmpty());
+    }
+
+    public function testToArray()
+    {
+        $context = new Context(['foo' => 'bar']);
+
+        $this->assertEquals(['foo' => 'bar'], $context->toArray());
     }
 
-    public function testMergeExtraDataWithSameKey()
+    public function testJsonSerialize()
+    {
+        $context = new Context(['foo' => 'bar']);
+
+        $this->assertEquals('{"foo":"bar"}', json_encode($context));
```

Comment:
Use `assertSame`

Reply:
Why? `assertSame` checks with strict equality which I don't see any benefit in using here